### PR TITLE
Introduce recover_dynamodb_lock

### PR DESCRIPTION
**Pull request by mosyp**
See docs at https://github.com/delta-io/kafka-delta-ingest/pull/79/files#diff-537aa1eb75dd77e0cfc4b6c45bbe0ec80fce6e3c9c8dbf054e8552beb8d746eaR18-R35

### CODE REVIEW

File: src/recover_dynamodb_lock.rs [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+                    return Ok(());
+                } else {
+                    log::error!("Could not delete the lock, it could be either deleted manually or by other writers. Trying again.");
+                }
+            }
+            _ => {
+                unreachable!()
+            }
+        }
+    }
```
Comment (mosyp) [inline src/recover_dynamodb_lock.rs:88]: The core changes

File: src/recover_dynamodb_lock.rs [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+/// In order to recover from this lock, KDI ensures that it is released first and waits at least
+/// the `min_wait_period` which is usually the write interval (`allowed_latency`). The additional
+/// pause before deletion is required to ensure that other writers will face the non-acquirable lock
+/// as well, which should happen in between `min_wait_period` interval.
+///
+/// The possible outcome if the lock is "recovered" by at least one worker was not failed before,
+/// could be that the `is_non_acquirable` lock was acquired to perform drop or drop/create scenario.
+/// If table is dropped but worker is still active, it'll keep writing despite there's no table;
+/// when table is recreated, the worker will keep writer with N+1 version, however the table is on
+/// version 0, since the delta log is cleared on drop.
```
Comment (xianwill) [inline src/recover_dynamodb_lock.rs:35]: :heart:

**Review by xianwill — APPROVED** [review]

**Review by houqp — COMMENTED** [review]
I am cool with us merge this as is since it's strictly better than what we have right now :)

But it looks like we still have some edge-cases that require more thoughts. So I left some comments in line.

File: src/recover_dynamodb_lock.rs [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+/// writers, which is the non-acquirable lock. In order to create one, the owner has to set the
+/// `is_non_acquirable` flag once acquire the lock. Other writers will fail to to try to acquire
+/// the log with this flag.
+///
+/// In order to recover from this lock, KDI ensures that it is released first and waits at least
+/// the `min_wait_period` which is usually the write interval (`allowed_latency`). The additional
+/// pause before deletion is required to ensure that other writers will face the non-acquirable lock
+/// as well, which should happen in between `min_wait_period` interval.
+///
+/// The possible outcome if the lock is "recovered" by at least one worker was not failed before,
```
[the remark this answers is not in the document]
Comment (mosyp) [reply src/recover_dynamodb_lock.rs:31]: Do you mean a table in this case? There is, but the things is that there's **update** call and then **rename**. And if the table drop/create happens in between, we screwed. That's why this new flag will make rename fail in a first place

File: src/recover_dynamodb_lock.rs [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+                return Ok(());
+            }
+            Some(item) if !item.is_non_acquirable => {
+                // lock is present but it's acquirable
+                log::info!("There is active lock but it is acquirable, no recover required.");
+                return Ok(());
+            }
+            Some(item) if item.is_non_acquirable && !item.is_released => {
+                // lock is present and non-acquirable but not yet released, let's wait
+                log::error!("There is active non-acquirable lock found, but it is not released yet, waiting...");
```
[the remark this answers is not in the document]
Comment (mosyp) [reply src/recover_dynamodb_lock.rs:66]: This is expected to happen tho, e.g. the non-acquirable lock holder still doing its business and lock is not released, but maybe it makes sense to move it to warn. I just figure that we catch any possible interaction with non-acquirable lock so it's not missed

File: src/recover_dynamodb_lock.rs [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+/// By default dynamodb lock once acquired could be either released by its owner or expired by
+/// other writers. However there's a type of locks that could not be expired or acquired by other
+/// writers, which is the non-acquirable lock. In order to create one, the owner has to set the
+/// `is_non_acquirable` flag once acquire the lock. Other writers will fail to to try to acquire
+/// the log with this flag.
+///
+/// In order to recover from this lock, KDI ensures that it is released first and waits at least
+/// the `min_wait_period` which is usually the write interval (`allowed_latency`). The additional
+/// pause before deletion is required to ensure that other writers will face the non-acquirable lock
+/// as well, which should happen in between `min_wait_period` interval.
```
[the remark this answers is not in the document]
Comment (mosyp) [reply src/recover_dynamodb_lock.rs:29]: yeah, we could make it 2x min_wait_period tho, or 1.5x, or even configurable. But yeah, I don't see yet any other option tbh

[the remark this answers is not in the document]
Comment (mosyp) [reply src/recover_dynamodb_lock.rs:29]: Ideally, if someone is trying to drop/create a table while KDI is running, he's better turn off any writers. The idea of this PR is to minimize the risk as much as possible

**Comment by mosyp** [comment]
Irrelevant to this PR. @xianwill this is the test error with default offset I was talking about, once in a while it could be reproduced even in ci https://github.com/delta-io/kafka-delta-ingest/runs/3672951682

e.g. the fix https://github.com/delta-io/kafka-delta-ingest/tree/default-offset

**Comment by mosyp** [comment]
As it turned out, this is only useful for drop/create scenario. And there's no 100% guarantee this to work with multiple writers, only best effort with timestamps. 

Closing this, as ppl should stop KDI before doing drop/create to ensure that delta store won't be corrupted

File: src/recover_dynamodb_lock.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (houqp) [reply src/recover_dynamodb_lock.rs:31]: yes, i meant table and being able to detect that at lock acquisition atomically. If we can do that, we can fail the rename if we detected that that lock is now only intended to be used for a newly recreated table.

**Pull request closed without merging**